### ISSUE TITLE
Semantics of unions and intersections of tables

### ISSUE THREAD

**Issue by taminomara** [EmmyLuaLs/emmylua-analyzer-rust#454]
This is a follow up to #447 and #449.

Currently, `A | B` is treated differently depending on context. In function calls, it is treated as "either type A or type B", leading to correct warnings:

```lua
--- @class A
--- @field a integer

--- @class B
--- @field b integer

--- @type A | B
local ab = {} -- correct warning: either `a` or `b` should be present

--- @type A | B
local ab = { a = 1 } -- correct absence of warning

--- @param x A
function bar(x) end

bar(ab) -- correct warning: `ab` can be `B`
```

However in method access, it is treated as "union of fields from type A and type B":

```lua
--- @class A
--- @field a integer

--- @class B
--- @field b integer

--- @type A | B
local ab = { a = 1 } -- correct absence of warning

print(ab.b) -- incorrect: no warning, even though `ab` can be `A`, which doesn't have `b`
```

All type inference should follow the first definition for union: `A | B` is "A or B", meaning that attribute access should generate a warning unless field is present in both `A` and `B`.

With this behavior, it will be possible to implement smarter type guards in the future:

```lua
--- @class A
--- @field d "a"
--- @field a integer

--- @class B
--- @field d "b"
--- @field b integer

--- @type A | B
local ab = { d = "a", a = 0 }

if ab.d == "a" then
    print(ab.a) -- should be ok: we've narrowed down type of `ab` to `A`.
    print(ab.b) -- should give warning: `A` doesn't have field `b`.
end
```

Intersection, on the other hand, should mean "union of fields from type A and type B":

```lua
--- @class A
--- @field a integer

--- @class B
--- @field b integer

--- @type A & B
local foo = { a = 1 } -- incorrect: no warning, even though field `b` is missing

--- @param x A
function bar(x) end

bar(foo) -- incorrect warning: `foo` can accept `A`, which is a supertype of `A & B`
```

For types that don't have a common supertype, intersections generally don't make sense:

```lua
--- @type string & integer
local x -- `x` can't string and integer at the same time.
```

Other languages deduce types of such intersections as `never`. Maybe we need to create such type, or use `void` in these cases.

For intersections of functions, typescript deduces them as overloads, but it's not strictly necessary:

```lua
--- @type (fun(x: integer)) & (fun(x: string))
local x

x(10) -- ok: `x` can accept an integer.
x("") -- ok: `x` can accept a string.
x({}) -- not ok: `x` can't accept a table.
x() -- not ok: `x` requires one argument.
```

Once again, this is a fairly niche feature, I don't think there will be much use for it.

**Comment by MillhioreBT** [comment]
> It is unlikely for us to support the precise semantics of `|` and `&` in TypeScript, considering the numerous challenges in our analysis:
> 
> 1. When we encounter type A, we cannot determine its members because class members can be defined across multiple files, which is not the case in TypeScript. As a result, we cannot determine what type `A & B` will yield during analysis.
> 2. Field-based type guards are also very challenging because I cannot determine the type of a variable during flow analysis. This variable might have a type declared using `@type`, or it might only be inferred. However, such inference requires analyzing other files to obtain the result, and I cannot make assumptions based on `require`.

It would be much better to have these types of well-defined diagnostics, no matter how dynamic it may be. There are specific cases like this example where the types and fields of each table are known beforehand, covering this case is the most important.

If someone defines with `---@type` or adds new values in other files, it should be our responsibility to deal with it, but the basic functionality of the `&` operator is to merge the fields of each table and generate a suitable generic type for the diagnostic.
